File: mobile/library/kotlin/io/envoyproxy/envoymobile/FinalStreamIntel.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.types.EnvoyFinalStreamIntel
import io.envoyproxy.envoymobile.engine.types.EnvoyStreamIntel

/**
 * Exposes one time HTTP stream metrics, context, and other details.
 *
 * Note: a timestamp field (ends with "Ms") with a value of -1 indicates that it is absent.
 *
 * @param streamId The stream identifier.
 * @param connectionId The connection identifier.
 * @param attemptCount The number of attempts used to perform a given request.
 * @param streamStartMs The time the stream started (a.k.a. request started), in ms since the epoch.
 * @param dnsStartMs The time the DNS resolution for this request started, in ms since the epoch.
 * @param dnsEndMs The time the DNS resolution for this request completed, in ms since the epoch.
 * @param connectStartMs The time the upstream connection started, in ms since the epoch. This may
 *   not be set if socketReused is false.
 * @param connectEndMs The time the upstream connection completed, in ms since the epoch. This may
 *   not be set if socketReused is false.
 * @param sslStartMs The time the SSL handshake started, in ms since the epoch. This may not be set
 *   if socketReused is false.
 * @param sslEndMs The time the SSL handshake completed, in ms since the epoch. This may not be set
 *   if socketReused is false.
 * @param sendingStartMs The time the first byte of the request was sent upstream, in ms since the
 *   epoch.
 * @param sendingEndMs The time the last byte of the request was sent upstream, in ms since the
 *   epoch.
 * @param responseStartMs The time the first byte of the response was received, in ms since the
 *   epoch.
 * @param streamEndMs The time when the stream reached a final state (Error, Cancel, Success), in ms
 *   since the epoch.
 * @param socketReused True if the upstream socket had been used previously.
 * @param sentByteCount The number of bytes sent upstream.
 * @param receivedByteCount The number of bytes received from upstream.
 * @param responseFlags The response flags for the stream.
 */
@Suppress("LongParameterList")
class FinalStreamIntel(
  streamId: Long,
  connectionId: Long,
  attemptCount: Long,
  val streamStartMs: Long,
  val dnsStartMs: Long,
  val dnsEndMs: Long,
  val connectStartMs: Long,
  val connectEndMs: Long,
  val sslStartMs: Long,
  val sslEndMs: Long,
  val sendingStartMs: Long,
  val sendingEndMs: Long,
  val responseStartMs: Long,
  val streamEndMs: Long,
  val socketReused: Boolean,
  val sentByteCount: Long,
  val receivedByteCount: Long,
  val responseFlags: Long
) : StreamIntel(streamId, connectionId, attemptCount) {
  constructor(
    superBase: EnvoyStreamIntel,
    base: EnvoyFinalStreamIntel
  ) : this(
    superBase.streamId,
    superBase.connectionId,
    superBase.attemptCount,
    base.streamStartMs,
    base.dnsStartMs,
    base.dnsEndMs,
    base.connectStartMs,
    base.connectEndMs,
    base.sslStartMs,
    base.sslEndMs,
    base.sendingStartMs,
    base.sendingEndMs,
    base.responseStartMs,
    base.streamEndMs,
    base.socketReused,
    base.sentByteCount,
    base.receivedByteCount,
    base.responseFlags
  )
}

File: mobile/library/swift/FinalStreamIntel.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// Exposes one time HTTP stream metrics, context, and other details.
/// Note: -1 means "not present" for the fields of type Int64.
@objcMembers
public final class FinalStreamIntel: StreamIntel {
  /// The time the stream started (a.k.a. request started), in ms since the epoch.
  public let streamStartMs: Int64
  /// The time the DNS resolution for this request started, in ms since the epoch.
  public let dnsStartMs: Int64
  /// The time the DNS resolution for this request completed, in ms since the epoch.
  public let dnsEndMs: Int64
  /// The time the upstream connection started, in ms since the epoch. (1)
  public let connectStartMs: Int64
  /// The time the upstream connection completed, in ms since the epoch. (1)
  public let connectEndMs: Int64
  /// The time the SSL handshake started, in ms since the epoch. (1)
  public let sslStartMs: Int64
  /// The time the SSL handshake completed, in ms since the epoch. (1)
  public let sslEndMs: Int64
  /// The time the first byte of the request was sent upstream, in ms since the epoch.
  public let sendingStartMs: Int64
  /// The time the last byte of the request was sent upstream, in ms since the epoch.
  public let sendingEndMs: Int64
  /// The time the first byte of the response was received, in ms since the epoch.
  public let responseStartMs: Int64
  /// The time the last byte of the request was received, in ms since the epoch.
  public let streamEndMs: Int64
  /// True if the upstream socket had been used previously.
  public let socketReused: Bool
  /// The number of bytes sent upstream.
  public let sentByteCount: UInt64
  /// The number of bytes received from upstream.
  public let receivedByteCount: UInt64
  /// The response flags for the upstream stream.
  public let responseFlags: UInt64
  /// The protocol of the upstream stream, or -1 if no stream was established.
  public let upstreamProtocol: Int64

  // NOTE(1): These fields may not be set if socket_reused is false.

  public init(
    streamId: Int64,
    connectionId: Int64,
    attemptCount: UInt64,
    streamStartMs: Int64,
    dnsStartMs: Int64,
    dnsEndMs: Int64,
    connectStartMs: Int64,
    connectEndMs: Int64,
    sslStartMs: Int64,
    sslEndMs: Int64,
    sendingStartMs: Int64,
    sendingEndMs: Int64,
    responseStartMs: Int64,
    streamEndMs: Int64,
    socketReused: Bool,
    sentByteCount: UInt64,
    receivedByteCount: UInt64,
    responseFlags: UInt64,
    upstreamProtocol: Int64
  ) {
    self.streamStartMs = streamStartMs
    self.dnsStartMs = dnsStartMs
    self.dnsEndMs = dnsEndMs
    self.connectStartMs = connectStartMs
    self.connectEndMs = connectEndMs
    self.sslStartMs = sslStartMs
    self.sslEndMs = sslEndMs
    self.sendingStartMs = sendingStartMs
    self.sendingEndMs = sendingEndMs
    self.responseStartMs = responseStartMs
    self.streamEndMs = streamEndMs
    self.socketReused = socketReused
    self.sentByteCount = sentByteCount
    self.receivedByteCount = receivedByteCount
    self.responseFlags = responseFlags
    self.upstreamProtocol = upstreamProtocol
    super.init(streamId: streamId, connectionId: connectionId, attemptCount: attemptCount)
  }
}

extension FinalStreamIntel {
  internal convenience init(_ cIntel: EnvoyStreamIntel, _ cFinalIntel: EnvoyFinalStreamIntel) {
    self.init(
      streamId: cIntel.stream_id,
      connectionId: cIntel.connection_id,
      attemptCount: cIntel.attempt_count,
      streamStartMs: cFinalIntel.stream_start_ms,
      dnsStartMs: cFinalIntel.dns_start_ms,
      dnsEndMs: cFinalIntel.dns_end_ms,
      connectStartMs: cFinalIntel.connect_start_ms,
      connectEndMs: cFinalIntel.connect_end_ms,
      sslStartMs: cFinalIntel.ssl_start_ms,
      sslEndMs: cFinalIntel.ssl_end_ms,
      sendingStartMs: cFinalIntel.sending_start_ms,
      sendingEndMs: cFinalIntel.sending_end_ms,
      responseStartMs: cFinalIntel.response_start_ms,
      streamEndMs: cFinalIntel.stream_end_ms,
      socketReused: cFinalIntel.socket_reused != 0,
      sentByteCount: cFinalIntel.sent_byte_count,
      receivedByteCount: cFinalIntel.received_byte_count,
      responseFlags: cFinalIntel.response_flags,
      upstreamProtocol: cFinalIntel.upstream_protocol
    )
  }
}
